### ISSUE TITLE
Duplicate declaration: Nginx::Resource::Location[/] is already declared;

### ISSUE THREAD

**Issue by slima** [voxpupuli/puppet-nginx#982]
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: agent 3.4.3, master 3.8.7
- Ruby: ruby 1.9.3p484
- Distribution: Ubuntu 14.04
- Module version: git master 2016-12-04

## How to reproduce (e.g Puppet code you use)

```
nginx::nginx_servers:
  'wiki.foo.pl':
    server_name       : ['wiki.foo.pl']
    ssl               : true
    ssl_redirect      : true
    ssl_port          : 443
    ssl_cert          : '/etc/ssl/foo/foo.pem'
    ssl_key           : '/etc/ssl/foo/foo.key'
    ssl_dhparam       : '/etc/ssl/dhparam.pem'
    www_root          : '/var/www/wiki/'
    index_files       : ['doku.php']
    use_default_location: false
    locations:
      '/':
        try_files         : ['$uri', '$uri/', '@dokuwiki']
        ssl               : true
        ssl_only          : true
      '~ \.php$':
        fastcgi           : '127.0.0.1:9000'
        fastcgi_split_path: '^(.+\.php)(/.+)$'
        try_files         : ['$uri', '=404']
        ssl               : true
        ssl_only          : true
      '@dokuwiki':
        rewrite_rules     :
          [
            '^/_media/(.*) /lib/exe/fetch.php?media=$1 last',
            '^/_detail/(.*) /lib/exe/detail.php?media=$1 last',
            '^/_export/([^/]+)/(.*) /doku.php?do=export_$1&id=$2 last',
            '^/(.*) /doku.php?id=$1&$args last'
          ]

  'wp.foo.pl':
    server_name       : ['wp.foo.pl']
    ssl               : true
    ssl_redirect      : true
    ssl_port          : 443
    ssl_cert          : '/etc/ssl/foo/foo.pem'
    ssl_key           : '/etc/ssl/foo/foo.key'
    ssl_dhparam       : '/etc/ssl/dhparam.pem'
    www_root          : '/var/www/wp/'
    index_files       : ['index.html','index.php']
    use_default_location: false
    locations:
      '/':
        try_files         : ['$uri', '$uri/', '/index.php?$args']
        ssl               : true
        ssl_only          : true
      '~ \.php$':
        fastcgi           : '127.0.0.1:9000'
        fastcgi_split_path: '^(.+\.php)(/.+)$'
        try_files         : ['$uri', '=404']
        ssl               : true
        ssl_only          : true
```

## What are you seeing

Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Duplicate declaration: Nginx::Resource::Location[/] is already declared; cannot redeclare at /etc/puppet/environments/development/modules/nginx/manifests/resource/server.pp:750 on node foo

## What behaviour did you expect instead

working?

**Comment by wyardley** [comment]
Yes, this should be better documented (and I'd love to find a fix that means people don't have to set this explicitly).
See #902 and #544. tldr; set `use_default_location` to `false`.

**Comment by slima** [comment]
I have set use_default_location to false. (See my code) One server works, two or more fails to compile.

**Comment by wyardley** [comment]
Sorry, my mistake -- missed that!

I didn't think location was unique globally, just per server block, let me see if I can reproduce this.

In the meantime, a workaround is just to use a unique title for each location, and set 'location' to '/'. I've had to do this in certain cases.

**Comment by wyardley** [comment]
Yes, I can definitely reproduce this, and in fact, I think we've got some ugly workarounds locally for this or a similar problem.

This manifest will also exhibit a duplicate resource error:
```

      class { 'nginx': }
      nginx::resource::server { 'www.puppetlabs.com':
        ensure   => present,
        www_root => '/var/www/www.puppetlabs.com',
        use_default_location => false,
        locations => { '/' =>
          {
            try_files => ['$uri', '$uri/', '@dokuwiki']
          }
        }
      }
      nginx::resource::server { 'www2.puppetlabs.com':
        ensure   => present,
        www_root => '/var/www/www.puppetlabs.com',
        use_default_location => false,
        locations => { '/' =>
          {
            try_files => ['index.html', 'index.html']
          }
        }
      }
```

Even though passing in locations under the server definition itself is definitely something that exists in the code, the example docs mostly / all seem to show using the syntax where locations and server blocks are defined separately from each other (in which case, it makes sense that the `title` of the location has to be unique, even if the `location` is not. I personally prefer the style you've used.

If you've got a good solution in the code, feel free to submit a PR. I'll see if I can ask around and see if there's a better way to fix this in the module's internals, but in the meantime, the workaround I suggested should fix it, e.g.,

```

      class { 'nginx': }
      nginx::resource::server { 'www.puppetlabs.com':
        ensure   => present,
        www_root => '/var/www/www.puppetlabs.com',
        use_default_location => false,
        locations => { 'www-root' =>
          {
            location => '/',
            try_files => ['$uri', '$uri/', '@dokuwiki']
          }
        }
      }
      nginx::resource::server { 'www2.puppetlabs.com':
        ensure   => present,
        www_root => '/var/www/www.puppetlabs.com',
        use_default_location => false,
        locations => { 'www2-root' =>
          {
            location  => '/',
            try_files => ['index.html', 'index.html']
          }
        }
      }
```

**Comment by wyardley** [comment]
ps - our workaround involved something like:
`create_resources('::nginx::resource::location', prefix($locations, $name), {`
however, since you can only use prefix on `name` and not `title` directly, I don't think there's a complete fix for this without still forcing someone to set `location` explicitly.

**Comment by jblom** [comment]
Trying to setup a config with uwsgi I also got the "Duplicate declaration" error. By adding the use_default_location = false on the server I at least get a good run using --noop (still have to test the whole thing on an empty box/VM)

(still have to see whether the whole uwsgi link will work, but at least the error mentioned here is gone)

This is my setup (with multiple locations):

       #create the virtual host
	nginx::resource::server { "${service_domain}" :
		ensure => present,
		listen_port => $service_port,
		use_default_location => false,
		server_name => ['localhost'],
		access_log => "/var/log/nginx/${service_name}_access.log",
		error_log => "/var/log/nginx/${service_name}_error.log",
	}

	#link the static files
	nginx::resource::location { '/static/' :
		ensure => present,
		server => "${service_domain}",
		location => '/static/',
		location_alias => "/home/${service_user}/${service_name}/src/static/",
	}

	#link to the uwsgi socket
	nginx::resource::location { '/' :
		ensure => present,
		server => "${service_domain}",
		location => '/',
		uwsgi => "unix:/tmp/${service_name}.sock",
		uwsgi_param => {
			'uwsgi_buffers' => '8 2M',
		},
	}

**Comment by jblom** [comment]
and thanks @wyardley of course for the suggest for using use_default_location = false

**Comment by plantroon** [comment]
I also ran into this and using use_default_location = false helped. It does not seem very logical/straight-forward. Thanks for the suggestion though.
